Update: (root) README

RafaelJohn9 opened the pull request:
## Changes

<!-- List the changes this PR makes. -->

- [x] Updated README

## Context

<!-- Explain why you're making these changes. -->

I’ve updated the `README.md` to improve the overall clarity and structure. The goal of these changes is to make it easier for new contributors and developers to understand how to set up the project locally, run tests, and begin contributing. I’ve tried to keep the original intent and flow intact, while rewording or reorganizing certain sections for better readability and guidance.

Thank you for your time and for reviewing the changes!

dscho commented on README.md:
```diff
@@ -1,256 +1,280 @@
-## Local development setup
+---
 
-> [!NOTE]
```
I am pretty concerned about removing an important visual marker like this.

I am not saying that you have to reinstate it if the new look gets too broken up by it, but I am saying that simply stating that the site was rewritten is a tad too little to convince me that all of these broad changes that are smooshed into a single commit are actually good.

dscho commented on README.md:
````diff
@@ -1,256 +1,280 @@
+    npm install @playwright/test
+    ```
 
-Supported browsers include `firefox`, `chromium`, `webkit`, `chrome`. You can also simply download all of them using `npx playwright install` but please first note that they all weigh >100MB, so you might want to refrain from doing that. Side note: In GitHub Actions' hosted runners, Chrome comes pre-installed, and you might be able to use your own Chrome installation, too, if you have one.
````
Hmm. I'm not a fan of dropping this important information on the floor.

dscho reviewed (commented):
While a quick glance at [the resulting README](https://github.com/git/git-scm.com/blob/97582b8444061a6b1148a9f1fa60600638e5c05a/README.md) looks pleasant, I am quite concerned about the form of this PR.

Take for example the commit message:

> Update: (root) README

That's it. I would be much more convinced of these changes if the commit message followed the guidance in https://github.blog/2022-06-30-write-better-commits-build-better-projects/ to improve it, in particular with a strong focus on this part:

  | What you’re doing | Why you’re doing it
-- | -- | --
High-level (strategic) | **Intent** (what does this accomplish?) | **Context** (why does the code do what it does now?)
Low-level (tactical) | **Implementation** (what did you do to accomplish your goal?) | **Justification** (why is this change being made?)

And then there are many, many, _many_ changes, all squashed into a single commit, making it very, very easy to miss anything undesired.

Simple example: the ` ```console ` construct is supposed to show a shell session, including prompts. Yet all of those `$` prompts are removed. Hrmpf.

Other example, some of the more in-depth information is simply removed. Maybe an improvable idea?

Instead, a large set of changes should be structured into individual commits, each with a commit message that makes a good case for the introduced changes.

One such commit could be to add a table of contents. That should probably even be one of the first commits of this PR, if not _the_ first.

Then, if you still insist on removing those `$` prompts from the shell sessions, that would need to be its own commit (if only to make it easier to drop _that_ part and not having to drop all other changes at the same time).

Don't get me wrong, I like a lot what I am seeing in the result. It's just that the current form of the PR makes it really, really hard to improve it. And improving it still needs, from my perspective at least.

dscho commented:
I'd just like to state that I stopped at only two comments because I anticipate that it will be much easier to review and improve this PR once the changes are broken up into a proper set of commits, not because there is nothing more to comment on.

RafaelJohn9 commented:
Hey @dscho , I understand your concerns,  I'll redo this PR in the manner you have suggested.

dscho commented:
> Hey @dscho , I understand your concerns, I'll redo this PR in the manner you have suggested.

@RafaelJohn9 thank you.

BTW if I were you, I'd use the technique described in https://git-scm.com/docs/git-rebase#_splitting_commits to avoid _redoing_ the PR but instead simply splitting all the existing changes into appropriate commits. I find that `git checkout -p <original-huge- commit>` helps me with such tasks.

RafaelJohn9 commented:
thanks  :+1: ,

everyday  is a new learning day :sweat_smile:

dscho commented:
> everyday is a new learning day 😅

Totally!

RafaelJohn9 committed:
Refactor README intro for clarity and layout

Move badges below the title to improve visual hierarchy. Reword the opening
paragraph to make the purpose of the site clearer for first-time visitors.

RafaelJohn9 committed:
Add detailed Table of Contents for project documentation

Replace initial local development note with a comprehensive Table of Contents
Organize sections for easy navigation: Getting Started, Local Development, Testing, Content Updates, Contributing, etc.

RafaelJohn9 committed:
Add "Getting Started" section with recommended cloning method
Introduce "Cloning the Repository" guide
Recommend using scalar for efficient sparse cloning

RafaelJohn9 committed:
Remove $ from shell commands to improve copy-pasting

Make commands easier to copy and paste without errors

RafaelJohn9 committed:
Add manual sparse clone instructions and remove $ from commands
Add "Alternative: Manual Sparse Clone" section for users without scalar

Provide clean, prompt-free shell commands to improve copy-pasting

Enhance  usability of the cloning instructions

RafaelJohn9 committed:
Refine note on using sparse-checkout with existing full clones

Reformat note for better readability and emphasis
Clarify instructions for users with a full clone to focus on subsets
Improve tone and consistency in documentation style

RafaelJohn9 committed:
Add detailed directory structure section with descriptions

Provide brief explanations for each relevant directory
Improve clarity for developers navigating the project layout

RafaelJohn9 committed:
Revamp local development section with prerequisites and serving instructions

Add a new "Local Development" section with subsections for prerequisites and serving the site
Specify required tools: Hugo (extended v0.128.0+), Node.js, and Windows WSL recommendation
Improve instructions for verifying Hugo version and serving the site locally
Clarify note about pre-rendered directories and GitHub workflows

RafaelJohn9 committed:
Clarify usage of Hugo’s built-in server with ugly URLs disabled

Reword instructions to use Hugo’s built-in server.
Explain the need to disable “ugly URLs” for compatibility with GitHub Pages
Add a clear note defining “ugly URLs” and why serve-public.js emulates GitHub Pages behavior

RafaelJohn9 committed:
Add "Enabling Search Locally" section and clean up shell command formatting

Introduce instructions to build the site with search enabled using Hugo and Pagefind
Remove $ prompt symbols from shell commands to improve copy-pasting experience
Improve clarity and usability of local testing documentation

RafaelJohn9 committed:
Update Pagefind server instructions and add note on performance impact

Rephrase to highlight using Pagefind’s built-in server.
Remove $ from shell commands for easier copy-pasting
Add a clear note about slower performance and disabled live reload when using Pagefind

RafaelJohn9 committed:
Revise Testing section for clarity and improved formatting

Rewrite introduction to describe Playwright test suite and location clearly
Add “Running Tests” subsection with step-by-step instructions
Remove $ prompt symbols from shell commands to simplify copy-pasting
Improve overall structure and readability of the testing setup documentation

RafaelJohn9 committed:
Clarify Playwright setup and testing instructions

Reorganize the Playwright testing section to improve clarity and usability.
Adds headings, formatting, and notes on browser download sizes.
Also explains how to run tests against local changes and filter tests with regex.

RafaelJohn9 committed:
Refactor manual pages section for clarity and structure

Add clear headings under 'Content Updates' for installing Ruby dependencies
and generating manual pages. Reorganize steps to separate local and remote
build instructions, and format code blocks consistently for better readability.

RafaelJohn9 committed:
Clarify manual page generation instructions and add localized build steps

Structured manual page instructions into local and GitHub-based workflows.
Added steps for building localized manual pages using git-html-l10n.
Improved formatting and clarity for version-specific builds and prerequisites.

RafaelJohn9 committed:
Refactor instructions for downloads data and ProGit book updates

Renamed "Update the ProGit book" section to "ProGit Book" for clarity.
Added step-by-step formatting for prerequisites and script usage.
Clarified how to fetch and update book content from local repositories.
Updated download data instructions for improved readability.

RafaelJohn9 committed:
Improve ProGit book update instructions

Reorganized steps into clearly numbered list.
Added cloning instructions and clarified directory structure.
Documented alternative GitHub-based update method using API tokens.
Explained how to build the book for all available languages.

RafaelJohn9 committed:
Improve contribution instructions

Encouraged contributions with a more welcoming tone.
Recommended using scalar and sparse-checkout for efficiency.
Added guidance on making pull requests and opening issues for features.

RafaelJohn9 committed:
Add detailed instructions for adding new GUI clients

Added a new section with step-by-step guidance on contributing GUI client entries.
Clarified metadata fields required in the .md file.
Provided image naming and formatting requirements.
Ensured platform and ordering fields are clearly explained.

RafaelJohn9 committed:
Refactor and enhance Useful Links section

Promoted "Hugo", "Pagefind", "Lychee", and "Playwright" to top-level headings.
Reorganized and standardized the formatting of resource links.
Improved clarity and readability by aligning structure with other sections.

RafaelJohn9 committed:
Improve formatting and clarity of License section

Promoted license section to a top-level heading.
Consolidated license descriptions for source code and graphical assets.
Enhanced readability and link presentation.

RafaelJohn9 commented:
Hey @dscho,

what's your opinion on this PR ?

dscho commented on README.md:
```diff
@@ -1,256 +1,292 @@
+
+#### 1. Recommended Approach: Using Scalar
+
+We recommend using [`scalar`](https://git-scm.com/docs/scalar) for an efficient and focused clone. This allows you to work only on the parts of the repository relevant to your interests.
```
I am still [concerned](https://github.com/git/git-scm.com/pull/2010/files#r2101877637) about the removal of that `> [!NOTE]` marker. It is a rather important thing that this sentence talks about, which would likely be missed otherwise.

dscho commented on README.md:
````diff
@@ -37,9 +37,9 @@ Welcome to the repository for [git-scm.com](https://git-scm.com), the official w
 ```console
-$ scalar clone https://github.com/git/git-scm.com
-$ cd git-scm.com/src
-$ git sparse-checkout set layouts content static assets hugo.yml data script
+scalar clone https://github.com/git/git-scm.com
+cd git-scm.com/src
+git sparse-checkout set layouts content static assets hugo.yml data script
````
If you remove the `$`, then the `console` markup is wrong, it should be `sh` instead.

dscho commented on README.md:
```diff
@@ -53,8 +53,8 @@ git sparse-checkout set layouts content static assets hugo.yml data script
 
-> [!NOTE]
-> If you _already_ have a full clone and wish to accelerate development by focusing only on a small subset of the pages, you may want to run the `git sparse-checkout set [...]` command mentioned above.
+> **Note:**  
```
Please compare the visual of `> [!NOTE]` vs `**Note:**`. I find the former vastly better and would highly recommend to keep it.

dscho commented on README.md:
```diff
@@ -56,19 +56,15 @@ git reset --hard
+- **.github/**: Contains GitHub workflow configurations.
+- **external/book/**, **external/docs/**: Pre-rendered pages (do not edit directly).
 
-- To add new GUIs:
-  - data/
```
While you may find it clearer, the original intention of this section was to answer the question "Which directories do I need in my sparse checkout?". That's kind of gone, the reader now has to cobble together what they need from that admittedly neat-looking (but less helpful) bullet list.

dscho commented on README.md:
```diff
@@ -66,30 +66,30 @@ git reset --hard
 
 ---
 
-- To work on the GitHub workflows that perform the automated, scheduled pre-rendering:
-  - .github/
+## Local Development
```
This commit (289edb62496b4fbaf3d7bed445680f8fbb19789d) is inconsistent. It removes some description of directories that should have been part of 5481388b99ee2d34909e2c5ab42b872965f106e9. This reduces the clarity of this PR and makes it quite hard to review.

dscho commented on README.md:
```diff
@@ -100,14 +100,14 @@ HUGO_UGLYURLS=false hugo serve -w
 > **Note:**  
 > "Ugly URLs" refer to URLs ending with `.html` (e.g., `/about.html`). GitHub Pages prefers these for compatibility. The `serve-public.js` script emulates this behavior.
 
-Pro-Tip: Do this in a sparse checkout that excludes large parts of `content/`, to speed up the rendering time.
```
This pro-tip is now removed? The end result might be clearer, and at the same time worse because helpful hints have been removed without adequate replacement.

dscho commented on README.md:
```diff
@@ -100,14 +100,14 @@ HUGO_UGLYURLS=false hugo serve -w
 > **Note:**  
 > "Ugly URLs" refer to URLs ending with `.html` (e.g., `/about.html`). GitHub Pages prefers these for compatibility. The `serve-public.js` script emulates this behavior.
 
-Pro-Tip: Do this in a sparse checkout that excludes large parts of `content/`, to speed up the rendering time.
+### Enabling Search Locally
 
-To test the site locally _with_ the search enabled, run this instead:
+To test the site with search enabled:
```
Without stressing that this is happening locally, would readers not be confused into thinking that they test the site in place, i.e. at https://git-scm.com/?

dscho commented on README.md:
````diff
@@ -122,19 +122,17 @@ npx -y pagefind --site public --serve
-```
+### Running Tests
 
-Since Playwright uses headless versions of popular web browsers, you most likely need to install at least one of them, e.g. via:
````
This is another important piece of information that seems to be culled in the name of clarity. I do disagree that this is a good idea.

dscho commented on README.md:
````diff
@@ -162,19 +162,15 @@ You can also run specific tests matching a pattern using `-g <regex>` for more f
 
 ---
 
-This will populate the manual pages for all Git versions. You can also populate them only for a specific Git version (faster):
+## Content Updates
 
-```console
-$ version=v2.23.0
-$ REBUILD_DOC=$version ruby ./script/update-docs.rb /path/to/git/.git en
-```
+### Manual Pages
 
-Or you can populate the man pages from GitHub (much slower) like this:
+1. Install Ruby prerequisites:
 
-```console
-$ export GITHUB_API_TOKEN=github_personal_auth_token
-$ REBUILD_DOC=$version ruby ./script/update-docs.rb remote en  # specific version
-```
+    ```console
+    bundler install
+    ```
````
Is it possible that you simply split the diff at the boundaries of the new section? If you carefully look at the removed lines and what was added in their place, the diff is non-sensical. I would have expected original text to be replaced by improved text, but basically talking about the exact same thing. That's distinctly not the case here.

dscho commented on README.md:
````diff
@@ -172,27 +172,32 @@ You can also run specific tests matching a pattern using `-g <regex>` for more f
+    ```console
+    ruby ./script/update-docs.rb /path/to/git/.git en
+    ```
 
-Or you can do it from GitHub (much slower) like this:
+   To build for a specific version:
 
-```console
-$ export GITHUB_API_TOKEN=github_personal_auth_token
-$ REBUILD_DOC=$version ruby ./script/update-docs.rb remote l10n  # specific version
-```
+    ```console
+    version=v2.23.0
+    REBUILD_DOC=$version ruby ./script/update-docs.rb /path/to/git/.git en
+    ```
 
-## Update the `Downloads` pages
+   To build from GitHub (slower):
 
-Now you need to get the latest downloads for the downloads pages:
+    ```console
+    export GITHUB_API_TOKEN=github_personal_auth_token
+    REBUILD_DOC=$version ruby ./script/update-docs.rb remote en
+    ```
 
-```console
-$ ruby ./script/update-download-data.rb
-```
+   For localized man pages, use a local clone of [git-html-l10n](https://github.com/jnavila/git-html-l10n):
+
+    ```console
+    ruby ./script/update-docs.rb /path/to/git-html-l10n/.git l10n
+    REBUILD_DOC=$version ruby ./script/update-docs.rb /path/to/git-html-l10n/.git l10n
+    ```
````
Why is `update-download-data` replaced with `update-docs`? This diff makes no sense, it is highly unclear what was improved, and in what way, and whether it is an improvement altogether. For a PR that wants to improve the clarity of the document, the PR itself lacks quite a bit of clarity.

dscho commented on README.md:
````diff
@@ -199,22 +199,21 @@ You can also run specific tests matching a pattern using `-g <regex>` for more f
     REBUILD_DOC=$version ruby ./script/update-docs.rb /path/to/git-html-l10n/.git l10n
     ```
 
-## Update the ProGit book
+### Downloads Data
 
-First, you will have to get the necessary prerequisites:
+Update the downloads data for the site:
 
 ```console
-$ bundler install
+ruby ./script/update-download-data.rb
````
This diff makes no sense.

dscho commented on README.md:
```diff
@@ -286,8 +286,7 @@ To add a new GUI client to the [list of GUI clients](https://git-scm.com/downloa
 
 ---
 
-The source code for the site is licensed under the MIT license, which you can find in
-the MIT-LICENSE.txt file.
+## License
 
-All graphical assets are licensed under the
-[Creative Commons Attribution 3.0 Unported License](https://creativecommons.org/licenses/by/3.0/).
+The source code for this site is licensed under the MIT license (see `MIT-LICENSE.txt`).  
+All graphical assets are licensed under the [Creative Commons Attribution 3.0 Unported License](https://creativecommons.org/licenses/by/3.0/).
```
Removing that line break makes the diff less clear.

dscho reviewed (commented):
While there is a lot of good stuff in this PR, there is at least an equal amount of unfortunate stuff that I have a hard time to bring myself to like. It was quite strenuous to try to make sense of those diffs in the commits, as it seems to me that the overall huge, honking, unreviewable diff was merely split at certain hunks, without regard whether the retained hunks actually reflect what their commit messages tried to convey. For example, removing a hint to run one script with a hint to write another script is wrong, and the underlying problem is that the removed hint should have been in a totally different commit (namely the one that rephrased that part of the README).

In its current form, I am not willing to merge this PR.

Do note, though, that I am not the maintainer of this site. There are other people who could merge this PR if you want to convince them (a strategy I would totally condone).

RafaelJohn9 commented:
Hey @dscho , Thank you for your detailed review ,

I apologize if this PR seems like a mess and does not satisfy your standards. if its okay, I would like to close this for now and create issues while submitting PRs.   That would make it easier to review and make our discussion not so over the place and rather be concentrated into specific points.

dscho commented:
> if its okay, I would like to close this for now and create issues while submitting PRs.

Totally okay with me. But as I said, I am not the maintainer of this site (even if it seems that the onus of reviewing and merging PRs somehow falls mostly on me, I guess I wasn't up in the tree at the count of three).

RafaelJohn9 commented:
hey @dscho , 

you are who we work with :)  :+1:
